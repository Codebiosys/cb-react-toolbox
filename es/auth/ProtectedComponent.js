'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

var _lodash = require('lodash');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactBootstrap = require('react-bootstrap');

var _WithPermissions = require('./WithPermissions');

var _WithPermissions2 = _interopRequireDefault(_WithPermissions);

var _actions = require('./actions');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var PERMISSION_DENIED_MESSAGE = 'You do not have access to this application. Please contact your system ' + 'administrator if you believe you\'ve recieved this message incorrectly.';

var ProtectedComponent = function (_Component) {
  _inherits(ProtectedComponent, _Component);

  function ProtectedComponent() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, ProtectedComponent);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = ProtectedComponent.__proto__ || Object.getPrototypeOf(ProtectedComponent)).call.apply(_ref, [this].concat(args))), _this), _this.renderPermissionDenied = function () {
      return _react2.default.createElement(
        'div',
        { className: 'container-fluid', style: { marginTop: '20px' } },
        _react2.default.createElement(
          _reactBootstrap.Panel,
          { bsStyle: 'danger' },
          _react2.default.createElement(
            _reactBootstrap.Panel.Heading,
            null,
            'Permission Denied'
          ),
          _react2.default.createElement(
            _reactBootstrap.Panel.Body,
            null,
            PERMISSION_DENIED_MESSAGE,
            _react2.default.createElement(
              'p',
              null,
              _react2.default.createElement(
                'a',
                { href: _this.props.authEndpoint },
                'Try logging in with different credentials \u2192'
              )
            )
          )
        )
      );
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(ProtectedComponent, [{
    key: 'componentWillMount',
    value: function componentWillMount() {
      this.checkAuthentication();
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate() {
      this.checkAuthentication();
    }
  }, {
    key: 'checkAuthentication',
    value: function checkAuthentication() {
      if (!this.props.user) {
        this.props.authenticate();
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          user = _props.user,
          children = _props.children,
          requiredPermissions = _props.requiredPermissions;


      if (!user) {
        return null;
      }

      return _react2.default.createElement(
        _react2.default.Fragment,
        null,
        _react2.default.createElement(
          _WithPermissions2.default,
          { has: [].concat(_toConsumableArray(requiredPermissions)) },
          children
        ),
        _react2.default.createElement(
          _WithPermissions2.default,
          { invert: true, has: [].concat(_toConsumableArray(requiredPermissions)) },
          this.renderPermissionDenied()
        )
      );
    }
  }]);

  return ProtectedComponent;
}(_react.Component);

ProtectedComponent.propTypes = {
  children: _propTypes2.default.node.isRequired,
  user: _propTypes2.default.object,
  authenticate: _propTypes2.default.func.isRequired,
  authEndpoint: _propTypes2.default.string.isRequired,
  requiredPermissions: _propTypes2.default.arrayOf(_propTypes2.default.string).isRequired
};
ProtectedComponent.defaultProps = {
  token: null,
  user: null
};


var mapStateToProps = function mapStateToProps(state) {
  return {
    user: (0, _lodash.get)(state, 'auth.user', false)
  };
};

var mapDispatchToProps = function mapDispatchToProps(dispatch, ownProps) {
  return {
    authenticate: function authenticate() {
      dispatch((0, _actions.authenticate)(ownProps.authEndpoint, ownProps.userEndpoint));
    }
  };
};

/**
 * @typedef {Class} ProtectedComponent
 */

exports.default = (0, _reactRedux.connect)(mapStateToProps, mapDispatchToProps)(ProtectedComponent);