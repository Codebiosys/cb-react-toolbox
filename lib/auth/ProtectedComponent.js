'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _toConsumableArray2 = require('babel-runtime/helpers/toConsumableArray');

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

var _lodash = require('lodash');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactBootstrap = require('react-bootstrap');

var _WithPermissions = require('./WithPermissions');

var _WithPermissions2 = _interopRequireDefault(_WithPermissions);

var _actions = require('./actions');

var _sagas = require('./sagas');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var PERMISSION_DENIED_MESSAGE = 'You do not have access to this application. Please contact your system ' + 'administrator if you believe you\'ve recieved this message incorrectly.';

var ProtectedComponent = function (_Component) {
  (0, _inherits3.default)(ProtectedComponent, _Component);

  function ProtectedComponent() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, ProtectedComponent);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = ProtectedComponent.__proto__ || (0, _getPrototypeOf2.default)(ProtectedComponent)).call.apply(_ref, [this].concat(args))), _this), _this.renderPermissionDenied = function () {
      return _react2.default.createElement(
        'div',
        { className: 'container-fluid', style: { marginTop: '20px' } },
        _react2.default.createElement(
          _reactBootstrap.Panel,
          { bsStyle: 'danger' },
          _react2.default.createElement(
            _reactBootstrap.Panel.Heading,
            null,
            'Permission Denied'
          ),
          _react2.default.createElement(
            _reactBootstrap.Panel.Body,
            null,
            PERMISSION_DENIED_MESSAGE,
            _react2.default.createElement(
              'p',
              null,
              _react2.default.createElement(
                'a',
                { href: (0, _sagas.getAuthUrl)() },
                'Try logging in with different credentials \u2192'
              )
            )
          )
        )
      );
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(ProtectedComponent, [{
    key: 'componentWillMount',
    value: function componentWillMount() {
      this.checkAuthentication();
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate() {
      this.checkAuthentication();
    }
  }, {
    key: 'checkAuthentication',
    value: function checkAuthentication() {
      if (!this.props.user) {
        this.props.authenticate();
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          user = _props.user,
          children = _props.children,
          requiredPermissions = _props.requiredPermissions;


      if (!user) {
        return null;
      }

      return _react2.default.createElement(
        _react2.default.Fragment,
        null,
        _react2.default.createElement(
          _WithPermissions2.default,
          { has: [].concat((0, _toConsumableArray3.default)(requiredPermissions)) },
          children
        ),
        _react2.default.createElement(
          _WithPermissions2.default,
          { invert: true, has: [].concat((0, _toConsumableArray3.default)(requiredPermissions)) },
          this.renderPermissionDenied()
        )
      );
    }
  }]);
  return ProtectedComponent;
}(_react.Component);

ProtectedComponent.defaultProps = {
  token: null,
  user: null
};


var mapStateToProps = function mapStateToProps(state) {
  return {
    user: (0, _lodash.get)(state, 'auth.user', false)
  };
};

var mapDispatchToProps = function mapDispatchToProps(dispatch) {
  return {
    authenticate: function authenticate() {
      dispatch((0, _actions.authenticate)());
    }
  };
};

/**
 * @typedef {Class} ProtectedComponent
 */

exports.default = (0, _reactRedux.connect)(mapStateToProps, mapDispatchToProps)(ProtectedComponent);