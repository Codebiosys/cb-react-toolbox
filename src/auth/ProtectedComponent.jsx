import React, { Component } from 'react';
import { connect } from 'react-redux';
import { get } from 'lodash';
import PropTypes from 'prop-types';
import { Panel } from 'react-bootstrap';

import WithPermissions from './WithPermissions';
import { authenticate } from './actions';
import { getAuthUrl } from './sagas';

const PERMISSION_DENIED_MESSAGE = (
  'You do not have access to this application. Please contact your system '
  + 'administrator if you believe you\'ve recieved this message incorrectly.'
);

class ProtectedComponent extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    user: PropTypes.object,
    authenticate: PropTypes.func.isRequired,
    requiredPermissions: PropTypes.arrayOf(
      PropTypes.string,
    ).isRequired,
  }

  static defaultProps = {
    token: null,
    user: null,
  }

  componentWillMount() {
    this.checkAuthentication();
  }

  componentDidUpdate() {
    this.checkAuthentication();
  }

  checkAuthentication() {
    if (!this.props.user) {
      this.props.authenticate();
    }
  }

  renderPermissionDenied = () => (
    <div className="container-fluid" style={{ marginTop: '20px' }}>
      <Panel bsStyle="danger">
        <Panel.Heading>Permission Denied</Panel.Heading>
        <Panel.Body>
          {PERMISSION_DENIED_MESSAGE}
          <p><a href={getAuthUrl()}>
            Try logging in with different credentials &#8594;
          </a></p>
        </Panel.Body>
      </Panel>
    </div>
  );

  render() {
    const { user, children, requiredPermissions } = this.props;

    if (!user) {
      return null;
    }

    return (
      <React.Fragment>
        <WithPermissions has={[...requiredPermissions]}>
          {children}
        </WithPermissions>
        <WithPermissions invert has={[...requiredPermissions]}>
          {this.renderPermissionDenied()}
        </WithPermissions>
      </React.Fragment>
    );
  }
}

const mapStateToProps = state => ({
  user: get(state, 'auth.user', false),
});

const mapDispatchToProps = dispatch => ({
  authenticate: () => {
    dispatch(authenticate());
  },
});

/**
 * @typedef {Class} ProtectedComponent
 */

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ProtectedComponent);
