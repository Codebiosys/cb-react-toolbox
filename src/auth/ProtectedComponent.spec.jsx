import React from 'react';
import { mount } from 'enzyme';
import { createMockStore } from 'redux-test-utils';
import PropTypes from 'prop-types';

import ProtectedComponent from './ProtectedComponent';

const authenticate = jest.fn();
const children = (<div>children</div>);

describe('The Messenger notifier', () => {
  const defaultStore = {
    app: {
      auth: {
        authenticated: true,
      },
    },
  };
  const unAuthdefaultStore = {
    app: {
      auth: {
        authenticated: false,
      },
    },
  };
  let store;
  let wrapper;
  beforeEach(() => {
    store = createMockStore(defaultStore);
    wrapper = mount(
      <ProtectedComponent authenticate={authenticate}>{children}</ProtectedComponent>,
      { context: { store },
        childContextTypes: { store: PropTypes.object },
      },
    );
  });

  it('renders', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('renders to null if not authenticated', () => {
    store = createMockStore(unAuthdefaultStore);
    wrapper = mount(
      <ProtectedComponent authenticate={authenticate}>{children}</ProtectedComponent>,
      { context: { store },
        childContextTypes: { store: PropTypes.object },
      },
    );
    expect(wrapper).toMatchSnapshot();
  });
});
