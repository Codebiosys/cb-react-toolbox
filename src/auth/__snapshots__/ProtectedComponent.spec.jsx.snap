// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Messenger notifier renders 1`] = `
<Connect(ProtectedComponent)
  authenticate={[MockFunction]}
>
  <ProtectedComponent
    authenticate={[Function]}
    token={null}
    user={false}
  />
</Connect(ProtectedComponent)>
`;

exports[`The Messenger notifier renders to null if not authenticated 1`] = `
<Connect(ProtectedComponent)
  authenticate={[MockFunction]}
>
  <ProtectedComponent
    authenticate={[Function]}
    token={null}
    user={false}
  />
</Connect(ProtectedComponent)>
`;
